### PR TITLE
Async execution

### LINKED ISSUE #56: Async Executables
Currently `Executable<T*>.Execute` is synchronous void. We need `AsyncExecutable<T*>.ExecuteAsync(CancellationToken)` or similar

### LINKED ISSUE #58: No tests for Executable
Partial blocker to #56

### PR DESCRIPTION
**Pull request by herecydev**
Resolves #56 
Resolves #58 

This is one of the first times I've written some async stuff, so hopefully it's okay!

### CODE REVIEW

File: tests/Compose.Tests/TransitionTests.cs [hunk 1]
```diff
@@ -79,7 +79,7 @@ public void CanTransitionService()
 		}
 
 		[Fact]
-		public void CanTransitionSpecifcallyBoundService()
+		public void CanTransitionSpecificallyBoundService()
```
Comment (smudge202) [inline tests/Compose.Tests/TransitionTests.cs]: Thanks for the spellcheck :)

**Comment by smudge202** [comment]
Looks good. The only consideration I can think of, is the implications of `Task.Run`. Currently, the assumption is the `Execution` is never asynchronous itself, it is simply synchronous code executed asynchronously.

If I wanted an asynchronous execution, I would probably inherit `Executable<Task>` and expect the `ExecuteAsync` to behave correctly. However, we'd be wrapping the async in an async.

Making sense? Basically, consider the implications of the Execution being async, not sync which seems to be the current assumption.

**Comment by herecydev** [comment]
Can you perform a check for async-ness with that?

```
if action is async
then await action
else
await Task.Run(action)
```

**Comment by smudge202** [comment]
How about:

```c#
var asyncAction = action as Action<Task>;
return await asyncAction ?? Task.Run(action);
```

No idea if compiles

**Comment by herecydev** [comment]
Don't think that's possible. In either case I'm not sure it's too big of an issue.

**Comment by smudge202** [comment]
I think it *may* be a big issue. If you're async'ing workload for length IO, for all we know `Task.Run` will block the thread it's running on and release the calling thread.

If we were handling async correctly, there would only be the calling thread, and it could release.

**Comment by smudge202** [comment]
https://msdn.microsoft.com/en-us/library/hh195051(v=vs.110).aspx

^ confirming, Task.Run spawns a thread. 

> Queues the specified work to run on the ThreadPool and returns a task handle for that work.

Imagine the Executable just reads a file using Async. That doesn't need threads spawned when using the `async` keyword.

**Comment by herecydev** [comment]
I'm *really* not very good at this, but could we do this?

``` c#
public void OnExecute(Action invoke)
{
	Execution = invoke;
}
public void OnExecute(Func<Task> invoke)
{
	ExecutionAsync = invoke;
 }
```

And then something like...

``` c#
if //some logic
await ExectionAsync();
else
await Task.Run(Execution, cancellationToken);
```

**Comment by smudge202** [comment]
Sure, something like that should be good. 

If they call `Execute` despite registering `Func<Task>` just do `ExecutionAsync().Wait()`?

**Comment by herecydev** [comment]
Yep and throw if they call both `.on execute()`?

**Comment by smudge202** [comment]
I think previously I just overwrote the property as opposed to exception?

I'd be tempted to replicate that behaviour, but taking into account you may need to null the other property?

**Comment by herecydev** [comment]
@smudge202 How's that looking now? I've just done the stock executable as an example for now

**Comment by smudge202** [comment]
I appreciate the code was completely void of tests before you stuck your head in there, but any chance you can add tests for what happens when specifying multiple `OnExecute`s, and just confirmatory behaviour showing you can mismatch `OnExecute(Func<Task>)` with `Execute`, and `OnExecute(Action)` with `ExecuteAsync`?

**Comment by herecydev** [comment]
@smudge202 Okay, testing should be fine for that now. I've modified the func slightly so that the cancellation token is honored.

**Pull request merged**